### PR TITLE
Move IGCE code from Action to Script Include

### PR DESCRIPTION
**Pull request by zachclark-ccpo**
# Overview
Optimized and refactored IGCE and its associated Script Includes, moving all affected parts into the `Igce` Script Includes
``` javascript
var Igce = Class.create();
Igce.prototype = {

    acquisitionPackageSysId: null,

    initialize: function (acquisitionPackageSysId) {
        this.acquisitionPackageSysId = acquisitionPackageSysId;
    },

    getEstimatedObligation: function (filterFn) {
        return this.getPayload().templatePayload.periodsEstimate.map(function (estimate) {
            return estimate.periodLineItems;
        }).reduce(function (pre, cur) {
            return pre.concat(cur);
        }).reduce(function (result, item) {
            return result + (item.monthlyPrice * item.monthsInPeriod);
        }, 0);
    },

    getEstimatedBaseYearObligation: function () {
        return this.getPayload().templatePayload.periodsEstimate.filter(function (estimate) {
            return estimate.period.periodType == "BASE";
        }).map(function (estimate) {
            return estimate.periodLineItems;
        }).reduce(function (pre, cur) {
            return pre.concat(cur);
        }).reduce(function (result, item) {
            return result + (item.monthlyPrice * item.monthsInPeriod);
        }, 0);
    },
```
This is the existing code, written by @jeffsegal-ccpo.

## getPackagePeriods
``` javascript
    getPackagePeriods: function (base, options) {
        // This returns the periods in the IGCE schema for the Periods Estimate Object
        var mappedPeriods = [];
        if (base) {
            mappedPeriods.push({
                "period": {
                    "periodType": base.period_type.toString(),
                    "periodUnitCount": base.period_unit_count.toString(),
                    "periodUnit": base.period_unit.toString(),
                    "optionOrder": base.option_order.toString()
                },
                "periodLineItems": [],
            });
        }
        if (options) {
            getListRecords(options,
                ["period_type", "period_unit_count", "period_unit", "option_order"],
                "x_g_dis_atat_period")
                .forEach(function (period) {
                    mappedPeriods.push({
                        "period": {
                            "periodType": period.period_type.toString(),
                            "periodUnitCount": period.period_unit_count.toString(),
                            "periodUnit": period.period_unit.toString(),
                            "optionOrder": period.option_order.toString()
                        }, "periodLineItems": []
                    });
                });
        }
        return mappedPeriods;
    },
```
This replaces the old usage of `mapPackagePeriods` , which was written for the DOW.
The old implementation used `mapPackagePeriods` and then went back through to format it to fit the IGCE style, this new function directly formats it to IGCE style, with no need to rely on external functions.

## getInstanceInfo
``` javascript
    getInstanceInfo: function (instance) {
        //This function retrieves:
        // classification_level
        // selected_periods
        // need_for_entire_task_order_duration
        // Appending it to the given Classification Instance, Environment Instance, Estimated Environment Instance

        var selectedPeriodsFields = ["period_type", "option_order", "period_unit", "period_unit_count"];

        var entireDuration = convertToBoolean(instance.need_for_entire_task_order_duration.toString());
        instance.need_for_entire_task_order_duration = entireDuration;

        if (entireDuration) {
            instance.selected_periods = [];
        } else {
            var selectedPeriods = [];
            var retrieved_periods = getListRecords(instance.selected_periods.toString(), selectedPeriodsFields, "x_g_dis_atat_period").map(removeSysId);
            for (var i = 0; i < retrieved_periods.length; i++) {
                var formatted_option_order;
                if (retrieved_periods[i].option_order == null) {
                    formatted_option_order = "";
                }
                else {
                    formatted_option_order = retrieved_periods[i].option_order.toString();
                }
                selectedPeriods.push({
                    "periodType": retrieved_periods[i].period_type.toString(),
                    "periodUnitCount": retrieved_periods[i].period_unit_count.toString(),
                    "periodUnit": retrieved_periods[i].period_unit.toString(),
                    "optionOrder": formatted_option_order
                });
            };
            instance.selected_periods = selectedPeriods;
        }

        var classLevel = new global.GlideQuery("x_g_dis_atat_classification_level")
            .get(instance.classification_level.toString(),
                ["impact_level", "classification"])
            .orElse(null);

        classLevel = removeSysId(classLevel);
        instance.classification_level = classLevel;
        return instance;
    },
```
This function is meant to replace the functionality of `getClassificationInstanceInfo`.
`getClassificationInstanceInfo` was reliant on: `outputs.package_periods_by_id[id]; ` , with `outputs` being an Action specific variable AND reliant on the old implementation of `mapPackagePeriods`.
This implementation isn't reliant on `outputs`, and was rewritten to exclude `mapPackagePeriods`.

## getNestedIgceServiceInfo
```javascript
    getNestedIgceServiceInfo: function (selectedService) {
        // Get Service Offering or user created offering for each selection
        var offering;
        if (selectedService.service_offering) {
            offering = new global.GlideQuery("x_g_dis_atat_service_offering")
                .get(selectedService.service_offering,
                    ["service_offering_group", "name"])
                .orElse(null);
            offering = removeSysId(offering);
        } else {
            offering = {
                name: selectedService.other_service_offering.toString(),
                service_offering_group: null
            };
        }
        selectedService.service_offering = offering;

        // Get Classification Instances info
        if (selectedService.classification_instances) {
            var classificationInstanceFields = [
                "need_for_entire_task_order_duration", "usage_description",
                "classification_level", "selected_periods", "monthly_price", "dow_task_number"];
            selectedService.classification_instances = getListRecords(
                selectedService.classification_instances.toString(),
                classificationInstanceFields,
                "x_g_dis_atat_classification_instance").map(removeSysId).map(Igce.prototype.getInstanceInfo);

        }

        // Get Estimated Environment Instances
        if (selectedService.estimated_environment_instances) {
            var estimatedEnvironmentInstanceFields = [
                "need_for_entire_task_order_duration", "classification_level", "selected_periods", "monthly_price",
                "dow_task_number", "instance_name", "instance_location", "number_of_vcpus", "storage_amount",
                "storage_type", "memory_amount", "memory_unit", "data_egress_monthly_amount", "data_egress_monthly_unit"
            ];
            selectedService.estimated_environment_instances = getListRecords(
                selectedService.estimated_environment_instances.toString(),
                estimatedEnvironmentInstanceFields,
                "x_g_dis_atat_estimated_environment_instance").map(removeSysId).map(Igce.prototype.getInstanceInfo);
        }
        return selectedService;
    },
```

This script was moved from Script Includes into Igce, and it also required a small refactor to appropriately return the desired output. Because of the change to using `getInstanceInfo`, this script was refactored to include it and properly map.

## getSelectedServiceOfferings

```javascript
    getSelectedServiceOfferings: function (selectedServiceOfferings) {
        // Retrieve selected service offerings, classification instances, and estimated environment instances 
        var selectedServiceFields = [
            "other_service_offering", "service_offering",
            "classification_instances", "estimated_environment_instances"
        ];

        var selected_service_offerings = getListRecords(
            selectedServiceOfferings.toString(),
            selectedServiceFields,
            "x_g_dis_atat_selected_service_offering")
            .map(removeSysId)
            .map(this.getNestedIgceServiceInfo);

        return selected_service_offerings;
    },

```
This is a brand new script, and turns functionality originally included in the Gather IGCE Data Action into a script that retrieves the specific `selectedServiceOfferrings` Array, turned into a function so it can be reused.
```javascript
    getCurrentEnvironment: function (currentEnvironment) {
        // Retrieve Environment Instance Fields
        var environmentInstanceFields = [
            "need_for_entire_task_order_duration", "instance_name",
            "classification_level", "selected_periods"];
        var env = new GlideRecord("x_g_dis_atat_current_environment");
        env.get(currentEnvironment.toString());

        var current_environment = getListRecords(
            env.environment_instances.toString(),
            environmentInstanceFields,
            "x_g_dis_atat_environment_instance").map(removeSysId).map(Igce.prototype.getInstanceInfo);
        return current_environment;
    },

```

This is a brand new script, and turns functionality originally included in the Gather IGCE Data Action into a script that retrieves the specific `currentEnvironment` of the Acquisition Package, turned into a function so it can be reused.

## mapEnvironmentInstance

```javascript
    mapEnvironmentInstance: function (data_to_map, periods_estimate, formatted_idiq_clins, is_current_env, service_offering) {
        // This works for Environment Instance or Estimated Environment Instance
        // For each Environment Instance in the data_to_map array
        data_to_map.forEach(function (item) {
            // Create environment configuration summary
            var configSummary = item.instance_location + " " + item.number_of_vcpus + " vCPU(s) "
                + item.storage_amount + " " + item.storage_type + " Storage "
                + item.memory_amount + " " + item.memory_unit + " Memory"
                + item.data_egress_monthly_unit + " " + item.data_egress_monthly_amount + " Data Egress."


            var serviceOffering, monthlyPrice, dowTaskNumber;
            // If its the current environment...
            if (is_current_env) {
                dowTaskNumber = "4.1.1";
                monthlyPrice = "N/A";
                serviceOffering = "Current Environment - " + item.instance_name;

            }
            else {
                dowTaskNumber = item.dow_task_number;
                serviceOffering = service_offering;
                monthlyPrice = item.monthly_price;
            }

            // If the Environment Instance is NOT needed for the entire TO duration
            // Map it to the periods specified in period_line_items
            if (!item.need_for_entire_task_order_duration) {

                // For each period in the selected_periods array of the estimated environment instance
                item.selected_periods.forEach(function (period) {

                    // Map period_type and base_period of the periods_estimate array
                    var formattedPeriodOutput = periods_estimate.map(function (input) {
                        return input.period.periodType + input.period.optionOrder;
                    });


                    var indexOfCurrentPeriod = formattedPeriodOutput.indexOf(period.periodType + period.optionOrder);


                    // Insert into periods_estimate, under the specified period and period_line_items /////////////////////Fix this output
                    periods_estimate[indexOfCurrentPeriod].periodLineItems.push({
                        "clin": "1000",
                        "idiqClin": getClinFromDowTask(item.dow_task_number, formatted_idiq_clins),
                        "dowTaskNumber": dowTaskNumber,
                        "serviceOffering": serviceOffering,
                        "itemDescriptionOrConfigSummary": configSummary,
                        "monthlyPrice": monthlyPrice,
                        "monthsInPeriod": monthsInPeriod(period.periodUnit, period.periodUnitCount),
                    });
                });
            }
            // Else if the Current Environment is needed for the entire TO duration
            // Map it to each period in the periods_estimate array
            else if (item.need_for_entire_task_order_duration) {
                periods_estimate.forEach(function (period) {
                    period.periodLineItems.push({
                        "clin": "x001",
                        "idiqClin": getClinFromDowTask(item.dow_task_number, formatted_idiq_clins),
                        "dowTaskNumber": dowTaskNumber,
                        "serviceOffering": serviceOffering,
                        "itemDescriptionOrConfigSummary": configSummary,
                        "monthlyPrice": monthlyPrice,
                        "monthsInPeriod": monthsInPeriod(period.period.periodUnit, period.period.periodUnitCount),
                    });
                });
            }
        });
    },
```
This function replaces the copy-pasta functionality that was present in the original Gather IGCE Data Action. It will map the Current Environment AND Estimated Environment Instance.

## mapClassificationInstance
```javascript
    mapClassificationInstance: function (data_to_map, periods_estimate, formatted_idiq_clins, service_offering) {
        data_to_map.forEach(function (instance) { ////FOR EACH CLASSIFICATION INSTANCE
            // Check if need for entire task order duration is false
            // If false, use the selected_periods array of the classification instance
            if (!instance.need_for_entire_task_order_duration) {
                // For each period in the selected_periods array of the classification instance
                instance.selected_periods.forEach(function (period) {
                    // Map period_type and base_period of the periods_estimate array
                    var formattedPeriodOutput = periods_estimate.map(function (input) {
                        return input.period.periodType + input.period.optionOrder;
                    });
                    var indexOfCurrentPeriod = formattedPeriodOutput.indexOf(period.periodType + period.optionOrder);
                    // Insert into periods_estimate, under the specified period and period_line_items
                    periods_estimate[indexOfCurrentPeriod].periodLineItems.push({
                        "clin": "x001",
                        "idiqClin": getClinFromDowTask(instance.dow_task_number, formatted_idiq_clins),
                        "dowTaskNumber": instance.dow_task_number,
                        "serviceOffering": service_offering.name,
                        "itemDescriptionOrConfigSummary": instance.usage_description,
                        "monthlyPrice": instance.monthly_price,
                        "monthsInPeriod": monthsInPeriod(period.periodUnit, period.periodUnitCount),

                    });
                });
            }
            // If instance is needed for the entire task_order_duration, add it to all period_line_items
            else if (instance.need_for_entire_task_order_duration) {
                periods_estimate.forEach(function (period) {
                    period.periodLineItems.push({
                        "clin": "x001",
                        "idiqClin": getClinFromDowTask(instance.dow_task_number, formatted_idiq_clins),
                        "dowTaskNumber": instance.dow_task_number,
                        "serviceOffering": service_offering.name,
                        "itemDescriptionOrConfigSummary": instance.usage_description,
                        "monthlyPrice": instance.monthly_price,
                        "monthsInPeriod": monthsInPeriod(period.period.periodUnit, period.period.periodUnitCount),
                    });
                });
            }
        });
    },
```
This function maps the Classification Instances of an item to the Periods Estimate array output.

## getClinInfo
```javascript
   getClinInfo: function (idiqClins) {
        var idiqFields = ["contract", "idiq_clin", "description", "contract_type", "classification_level"];
        //put everything into an array
        var clin_info = getListRecords(idiqClins.toString(), idiqFields, 'x_g_dis_atat_idiq_clin').map(removeSysId);
        clin_info.forEach(function (item) {
            if (item.classification_level != null) {
                var gr = new GlideRecord('x_g_dis_atat_classification_level');
                gr.get(item.classification_level.toString());
                var classLevel = gr.getDisplayValue('classification');
                item.classification_level = classLevel.toUpperCase();
                // IGCE appends a CLASSFIED to the end of SECRET and TOP SECRET, format string to reflect that difference
                if (item.classification_level == "SECRET" || item.classification_level == "TOP SECRET") {
                    item.classification_level += " CLASSIFIED";
                }

            }
        });

        var formatted_clin_info = [];
        clin_info.forEach(function (item1) {
            clin_info.forEach(function (item2) {
                var idiqClinString;
                // If the items descriptions and classification_levels match, create a new idiqClinString
                if (item1.description == item2.description && item1.classification_level == item2.classification_level && item1.idiq_clin != item2.idiq_clin) {
                    // Lowest IDIQ CLIN first
                    if (parseInt(item1.idiq_clin) < parseInt(item2.idiq_clin)) {
                        idiqClinString = "x" + item1.idiq_clin.substr(1) + "/x" + item2.idiq_clin.substr(1) + " " + item1.description + " " + item1.classification_level;
                    }
                    else {
                        idiqClinString = "x" + item2.idiq_clin.substr(1) + "/x" + item1.idiq_clin.substr(1) + " " + item1.description + " " + item1.classification_level;
                    }
                    // If it doesn't exist in the array of new CLINs, push it
                    if (formatted_clin_info.indexOf(idiqClinString) == -1) {
                        formatted_clin_info.push(idiqClinString);
                    }
                }
                // If the description is Travel or External Agency Fee, this is a one-off IDIQ CLIN
                else if (item1.description == "Travel" || item1.description == "External Ordering Agency Fee") {
                    idiqClinString = "x" + item1.idiq_clin.substr(1) + " " + item1.description + " " + item1.contract_type;
                    if (formatted_clin_info.indexOf(idiqClinString) == -1) {
                        formatted_clin_info.push(idiqClinString);
                    }
                }
            });
        });
        return formatted_clin_info;
    },
```
This script is unchanged from the original IGCE Action, although I anticipate it will be reworked in the future (see Epic for updating the IGCE).

## getPayload
```javascript
    getPayload: function () {
        var acq_package = this.acquisitionPackageSysId;
        var periodOfPerformance = acq_package.period_of_performance;
        var currentEnvironment = acq_package.current_environment;
        var selectedServiceOfferings = acq_package.selected_service_offerings;
        var idiqClins = acq_package.idiq_clins;
        var fundingRequest = acq_package.funding_request;
        var requirementsCostEstimate = acq_package.requirements_cost_estimate;


        /////////////////////////////////////////////////////////////////////////////////////////////////////////////
        // TODO: 
        // In order to prevent SNOW Error: Cannot convert null to an object.,Detail: Cannot convert null to an object.
        // Check for missing fields, return an error if they don"t exist
        if (!periodOfPerformance) {
            // return an error
            return {
                error: "Missing Period of Performance"
            };
        }

        if (!selectedServiceOfferings && !currentEnvironment) {
            // return an error
            return {
                error: "Selected Service Offerings OR Current Environment are required to generate IGCE"
            };
        }


        ///////////////////////////////////////////////////////////////////////////////////////////////////////////////

        // Get Formatted CLINs
        var formattedIdiqClins = Igce.prototype.getClinInfo(idiqClins);

        // Get Periods for Period's Estimate
        var periodsEstimate = this.getPackagePeriods(
            periodOfPerformance.base_period, periodOfPerformance.option_periods
        );

        // If Current Environment Exists...
        if (currentEnvironment) {
            //map info to periodsEstimate
            var currentEnv = this.getCurrentEnvironment(currentEnvironment);
            var isCurrentEnv = true;
            Igce.prototype.mapEnvironmentInstance(currentEnv, periodsEstimate, formattedIdiqClins, isCurrentEnv);
        }

        // If Selected Service Offerings Exists (reference lists)
        if (selectedServiceOfferings) {
            //Map info to periodsEstimate
            var serviceOfferings = this.getSelectedServiceOfferings(selectedServiceOfferings);
            serviceOfferings.forEach(function (item) {
                if (item.classification_instances) {
                    var classificationInstances = item.classification_instances;
                    Igce.prototype.mapClassificationInstance(classificationInstances, periodsEstimate, formattedIdiqClins, item.service_offering);
                }
                else if (item.estimated_environment_instances) {
                    Igce.prototype.mapEnvironmentInstance(item, periodsEstimate, formattedIdiqClins, isCurrentEnv, item.service_offering);
                }
            });
        }

        // Get Funding Request
        var fundingDocument;
        if (fundingRequest) {
            if (fundingRequest.funding_request_type == "MIPR") {
                fundingDocument = {
                    "fundingType": fundingRequest.funding_request_type.toString(),
                    "miprNumber": fundingRequest.mipr.mipr_number.toString()
                }
            }
            else if (fundingRequest.funding_request_type == "FS_FORM") {
                fundingDocument = {
                    "fundingType": fundingRequest.funding_request_type.toString(),
                    "gtcNumber": fundingRequest.fs_form.gt_c_number.toString(),
                    "orderNumber": fundingRequest.fs_form.order_number.toString()
                }
            }
        }

        // Get Surge Capabilities
        var surgeCapabilities
        if (requirementsCostEstimate) {
            var gr = new GlideRecord('x_g_dis_atat_requirements_cost_estimate');
            gr.get(requirementsCostEstimate);
            surgeCapabilities = parseInt(gr.surge_capabilities.toString());
        }

        // IGCE generated-document payload
        var igce_payload = {
            documentType: "INDEPENDENT_GOVERNMENT_COST_ESTIMATE",
            templatePayload: {
                fundingDocument: fundingDocument,
                surgeCapabilities: surgeCapabilities,
                periodsEstimate: periodsEstimate,
            }
        };
        // Format the final payload
        return igce_payload;
    }
};
```
This function is the _clean_ implementation of the "Gather required IGCE Data" step of the original ATAT Gather IGCE Data action. It calls the 8 new functions implemented to gather the data required to return an `igce_payload` JSON object.

# Testing
## Validate ATAT Gather IGCE Data has been updated to use `.getPayload()`
image

## Use Test Action with an Acquisition Package Record
image

You will need to create a new Acquisition Package with the following fields:
Acquisition Package
- Period of Performance, with Base and Option Periods
- **OPTIONAL**: Current Environment
- Selected Service Offerings
- IDIQ CLINs
- Funding Request
- Requirements Cost Estimate

You can also check which fields are required by looking at the `getPayload` function:
```
   getPayload: function () {
        var acq_package = this.acquisitionPackageSysId;
        var periodOfPerformance = acq_package.period_of_performance;
        var currentEnvironment = acq_package.current_environment;
        var selectedServiceOfferings = acq_package.selected_service_offerings;
        var idiqClins = acq_package.idiq_clins;
        var fundingRequest = acq_package.funding_request;
        var requirementsCostEstimate = acq_package.requirements_cost_estimate;
```

## Important Testing Notes
Your Environment Instances / Classification Instances / Estimated Environment Instances are required to have **Need For Entire Task Order Duration** field filled out.
Also, the Periods on Environment Instances / Classification Instances / Estimated Environment Instances Selected Period's Field **MUST** be present in the original Acquisition Package's Period of Performance.

Ticket: AT-8110

### CODE REVIEW

**Comment by zachclark-ccpo** [comment]
10/26 Update:
image

Currently running into an issue where SNOW is claiming
```
Error: undefined is not a function.,Detail: undefined is not a function.
```

I have narrowed it down to `getNestedIgceServiceInfo`
Specifically the final line:
```
        // Get Classification Instances info
        if (selectedService.classification_instances) {
            var classificationInstanceFields = [
                'need_for_entire_task_order_duration', 'usage_description',
                'classification_level', 'selected_periods', 'monthly_price', 'dow_task_number'];
            selectedService.classification_instances = getListRecords(
                selectedService.classification_instances.toString(),
                classificationInstanceFields,
                'x_g_dis_atat_classification_instance').map(removeSysId).map(this.getInstanceInfo);
```
The function cannot recognize `getInstanceInfo`, this issue is likely caused by `this`.

**Comment by jasonburkert-ccpo** [comment]
Can you try the suggestion **Don't use this** from https://stackoverflow.com/a/20279485
I think it would work like this...
```
      // Get Classification Instances info
      if (selectedService.classification_instances) {
          var classificationInstanceFields = [
              'need_for_entire_task_order_duration', 'usage_description',
              'classification_level', 'selected_periods', 'monthly_price', 'dow_task_number'];
          var self = this;
          selectedService.classification_instances = getListRecords(
              selectedService.classification_instances.toString(),
              classificationInstanceFields,
              'x_g_dis_atat_classification_instance').map(removeSysId).map(self.getInstanceInfo);
      }
```


> 10/26 Update: image
> 
> Currently running into an issue where SNOW is claiming
> 
> ```
> Error: undefined is not a function.,Detail: undefined is not a function.
> ```
> 
> I have narrowed it down to `getNestedIgceServiceInfo` Specifically the final line:
> 
> ```
>         // Get Classification Instances info
>         if (selectedService.classification_instances) {
>             var classificationInstanceFields = [
>                 'need_for_entire_task_order_duration', 'usage_description',
>                 'classification_level', 'selected_periods', 'monthly_price', 'dow_task_number'];
>             selectedService.classification_instances = getListRecords(
>                 selectedService.classification_instances.toString(),
>                 classificationInstanceFields,
>                 'x_g_dis_atat_classification_instance').map(removeSysId).map(this.getInstanceInfo);
> ```
> 
> The function cannot recognize `getInstanceInfo`, this issue is likely caused by `this.`. Because `getNestedIgceServiceInfo` is itself being called by a `this.`, a nested `this` is causing the function to be undefined. I know this because I use `this.getInstanceInfo` in other places and it works perfectly.

**Comment by zachclark-ccpo** [comment]
> Can you try the suggestion **Don't use this** from https://stackoverflow.com/a/20279485 I think it would work like this...
> 
> ```
>       // Get Classification Instances info
>       if (selectedService.classification_instances) {
>           var classificationInstanceFields = [
>               'need_for_entire_task_order_duration', 'usage_description',
>               'classification_level', 'selected_periods', 'monthly_price', 'dow_task_number'];
>           var self = this;
>           selectedService.classification_instances = getListRecords(
>               selectedService.classification_instances.toString(),
>               classificationInstanceFields,
>               'x_g_dis_atat_classification_instance').map(removeSysId).map(self.getInstanceInfo);
>       }
> ```
> 
> > 10/26 Update: image
> > Currently running into an issue where SNOW is claiming
> > ```
> > Error: undefined is not a function.,Detail: undefined is not a function.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I have narrowed it down to `getNestedIgceServiceInfo` Specifically the final line:
> > ```
> >         // Get Classification Instances info
> >         if (selectedService.classification_instances) {
> >             var classificationInstanceFields = [
> >                 'need_for_entire_task_order_duration', 'usage_description',
> >                 'classification_level', 'selected_periods', 'monthly_price', 'dow_task_number'];
> >             selectedService.classification_instances = getListRecords(
> >                 selectedService.classification_instances.toString(),
> >                 classificationInstanceFields,
> >                 'x_g_dis_atat_classification_instance').map(removeSysId).map(this.getInstanceInfo);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The function cannot recognize `getInstanceInfo`, this issue is likely caused by `this.`. Because `getNestedIgceServiceInfo` is itself being called by a `this.`, a nested `this` is causing the function to be undefined. I know this because I use `this.getInstanceInfo` in other places and it works perfectly.

Resolved.
Had to change the call to:
```
.map(Igce.prototype.getInstanceInfo);
```
inside of `getNestedIgceServiceInfo`

File: f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml [hunk 1]
```diff
@@ -39,106 +39,426 @@ Igce.prototype = {
...
+        var currentEnvironment = acq_package.current_environment;
+        var selectedServiceOfferings = acq_package.selected_service_offerings;
+        var idiqClins = acq_package.idiq_clins;
+        var fundingRequest = acq_package.funding_request;
+        var requirementsCostEstimate = acq_package.requirements_cost_estimate;
 
 
-        return this.periods_estimate;
+        /////////////////////////////////////////////////////////////////////////////////////////////////////////////
+        // TODO:
```
Comment (juliusfitzhugh-ccpo) [inline f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml:361]: I like the sectioning off of the `TODO:`,  is the thought to add more here in the future?

File: f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml [hunk 2]
```diff
@@ -9,138 +9,456 @@
...
+        var formatted_clin_info = [];
+        clin_info.forEach(function (item1) {
+            clin_info.forEach(function (item2) {
+                var idiqClinString;
+                // If the items descriptions and classification_levels match, create a new idiqClinString
+                if (item1.description == item2.description && item1.classification_level == item2.classification_level && item1.idiq_clin != item2.idiq_clin) {
+                    // Lowest IDIQ CLIN first
+                    if (parseInt(item1.idiq_clin) < parseInt(item2.idiq_clin)) {
+                        idiqClinString = "x" + item1.idiq_clin.substr(1) + "/x" + item2.idiq_clin.substr(1) + " " + item1.description + " " + item1.classification_level;
+                    }
```
Comment (juliusfitzhugh-ccpo) [inline f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml:330]: Is two `forEach` loops needed on `clin_info`? I don't completely understand what is happening by having two. Could sorting prior to the `forEach` help?

**Review by juliusfitzhugh-ccpo — APPROVED** [review]
Nice refactoring, the breaking up of the different sections into methods help show what each section is doing and what data is being gathered. I know there were some bumps, but you made it through and improved sections through the refactor. This is a large amount of data to process, nice work.

I left a few comments, but nothing that is blocking. The intent of the ticket has been captured, and the IGCE data is gathered using a script includes.

File: f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml (hunk 1 above)
Comment (jasonburkert-ccpo) [reply f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml:361]: It seems like the intent of the TODO is to provide the null checks deeper in the data structure.  For now they're left at a high level.

File: f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml (hunk 2 above)
Comment (jasonburkert-ccpo) [reply f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml:330]: @zachclark-ccpo Can you provide some insight here?  It looks like you are comparing two CLINs to determine whether an IDIQ CLIN is needed.

**Comment by jasonburkert-ccpo** [comment]
@zachclark-ccpo Nice effort. Please avoid buses.

File: f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml (hunk 1 above)
Comment (zachclark-ccpo) [reply f600233d1b154d507b782f84604bcb12/update/sys_script_include_b29002b097a211106fa8b4b3f153afc4.xml:361]: Yeah I was hoping we could get a unified error response in the future when we rework the IGCE.

There is the `ErrorHandler` script include, however I went way out of scope for this ticket and spent too much time refactoring to spend any extra time setting that up with the `Igce` script.

**Pull request merged**